File: gripss/src/main/java/com/hartwig/hmftools/gripss/links/Link.java
package com.hartwig.hmftools.gripss.links;

import static java.lang.Math.abs;

import static com.hartwig.hmftools.common.utils.sv.StartEndIterator.SE_END;
import static com.hartwig.hmftools.common.utils.sv.StartEndIterator.SE_START;

import com.hartwig.hmftools.gripss.common.Breakend;
import com.hartwig.hmftools.gripss.common.SvData;

public class Link
{
    public final String Id;
    private final Breakend[] mBreakends;

    public static final String LINK_TYPE_PAIR = "PAIR"; // used a single SV, not pairs of breakends from different SVs

    public Link(final String id, final Breakend first, final Breakend second)
    {
        Id = id;
        mBreakends = new Breakend[] { first, second };
    }

    public static Link from(final SvData sv)
    {
        return new Link(LINK_TYPE_PAIR, sv.breakendStart(), sv.breakendEnd());
    }

    public static Link from(final String linkId, final Breakend first, final Breakend second)
    {
        return new Link(linkId, first, second);
    }

    public String vcfId() { return mBreakends[SE_START].VcfId; }
    public Breakend breakendStart() { return mBreakends[SE_START]; }
    public Breakend breakendEnd() { return mBreakends[SE_END]; }

    public Breakend otherBreakend(final Breakend breakend) { return breakendStart() == breakend ?  breakendEnd() : breakendStart(); }

    public int minDistance()
    {
        if(breakendStart().sv().equals(breakendEnd().sv()))
            return breakendStart().sv().duplicationLength() + breakendStart().sv().insertSequenceLength();

        return abs(breakendStart().maxPosition() - breakendEnd().minPosition());
    }

    public int maxDistance()
    {
        if(breakendStart().sv().equals(breakendEnd().sv()))
            return breakendStart().sv().duplicationLength() + breakendStart().sv().insertSequenceLength();

        return abs(breakendStart().minPosition() - breakendEnd().maxPosition());
    }

    public Link reverse()
    {
        return new Link(Id, mBreakends[SE_END], mBreakends[SE_START]);
    }

    public String toString() { return String.format("%s<%s>%s distance(%d - %d)",
            breakendStart().VcfId, Id, breakendEnd().VcfId, minDistance(), maxDistance()); }
}

File: gripss-kt/src/main/kotlin/com/hartwig/hmftools/gripsskt/link/Link.kt
package com.hartwig.hmftools.gripsskt.link

import com.hartwig.hmftools.gripsskt.StructuralVariantContext
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

data class Link(val link: String, val vcfId: String, val otherVcfId: String, val minDistance: Int, val maxDistance: Int) {

    companion object {
        operator fun invoke(variant: StructuralVariantContext): Link {
            val distance = variant.duplicationLength + variant.insertSequenceLength
            return Link("PAIR", variant.vcfId, variant.mateId!!, distance, distance)
        }

        operator fun invoke(link: String, variants: Pair<StructuralVariantContext, StructuralVariantContext>): Link {
            val (minDistance, maxDistance) = distance(variants.first, variants.second)
            return Link(link, variants.first.vcfId, variants.second.vcfId, minDistance, maxDistance)
        }


        private fun distance(first: StructuralVariantContext, second: StructuralVariantContext): Pair<Int, Int> {
            val minDistance = abs(first.maxStart - second.minStart)
            val maxDistance = abs(first.minStart - second.maxStart)
            return Pair(min(minDistance, maxDistance), max(minDistance, maxDistance))
        }
    }

    fun reverse(): Link {
        return Link(link, otherVcfId, vcfId, minDistance, maxDistance)
    }

    override fun toString(): String {
        return "$vcfId<$link>$otherVcfId"
    }
}